apps/emqx_resource/src/emqx_resource_cache.erl

```diff
@@ -0,0 +1,331 @@
+
+get_cb_pt(ID) ->
+    try
+        persistent_term:get(?CB_PT_KEY(ID))
+    catch
+        error:badarg ->
+            undefined
+    end.
+
+to_channel_record({ID0, #{status := Status, error := Error}}) ->
+    ID = split_channel_id(ID0),
+    #channel{
+        id = ID,
+        status = Status,
+        error = Error,
+        extra = []
+    }.
+
+split_channel_id(Id) when is_binary(Id) ->
+    case binary:split(Id, <<":">>, [global]) of
+        [
+            ChannelGlobalType,
+            ChannelSubType,
+            ChannelName,
+            <<"connector">>,
+            ConnectorType,
+            ConnectorName
+        ] ->
+            ConnectorId = <<"connector:", ConnectorType/binary, ":", ConnectorName/binary>>,
+            ChannelId =
+                <<ChannelGlobalType/binary, ":", ChannelSubType/binary, ":", ChannelName/binary>>,
+            {ConnectorId, ChannelId};
+        _ ->
+            %% this is not a per-channel query, e.g. for authn/authz
+            {Id, ?NO_CHANNEL}
+    end.
+
+%% State can be quite bloated, caching it in ets means excessive large term copies,
+%% for each and every query so we keep it in persistent_term instead.
+%% Connector state is relatively static, so persistent_term update triggered GC is less of a concern
+%% comparing to other fields such as `status' and `error', which may change very often.
+put_state_pt(ID, State) ->
+    case get_cb_pt(ID) of
+        S when S =:= State ->
+            %% identical
+            ok;
+        _ ->
+            _ = persistent_term:put(?CB_PT_KEY(ID), State),
+            ok
+    end.
+
+del_state_pt(ID) ->
+    _ = persistent_term:erase(?CB_PT_KEY(ID)),
+    ok.
+
+is_exist(ID) ->
+    ets:member(?CACHE, ID).
+
+make_resource_data(ID, Connector, Channels) ->
+    #connector{
+        st_err = #{
+            error := Error,
+            status := Status
+        },
+        config = Config,
+        cb = Cb0
+    } = Connector,
+    Cb =
+        case Cb0 of
+            undefined ->
+                get_cb_pt(ID);
+            X ->
+                X
+        end,
+    #{
+        mod := Mod,
+        callback_mode := CallbackMode,
+        query_mode := QueryMode,
+        state := State
+    } = Cb,
+    #{
+        id => ID,
+        mod => Mod,
+        callback_mode => CallbackMode,
+        query_mode => QueryMode,
+        error => Error,
+        status => Status,
+        config => Config,
+        added_channels => Channels,
+        state => State
+    }.
+
+find_channels(ConnectorID) ->
+    MS = ets:fun2ms(fun(#channel{id = {Cid, _}} = C) when Cid =:= ConnectorID -> C end),
+    List = ets:select(?CACHE, MS),
+    lists:foldl(
+        fun(
+            #channel{
+                id = {ConnectorId, ChannelId},
+                status = Status,
+                error = Error
+            },
+            Acc
+        ) ->
+            Key = iolist_to_binary([ChannelId, ":", ConnectorId]),
```

Comment:
Q: do we need to append the connector id here, since we're already searching for channels from a specific connector?
Maybe we could only construct it like this in `make_resource_data`, and `erase_channel` could work with tuples/iolists.

Reply:
maybe, but `erase_old_channels` takes the external format chain IDs.

Reply:
exactly, erase_old_channels could handle tuples/iolists.

Reply:
I mean this is externalized 
https://github.com/emqx/emqx/pull/13838/files#diff-67a1a968ee134d34263236a53df607c87f2688b085300b25dee20b423b9f2e5dR104